### ISSUE TITLE
[metals] Diagonistics don't update in realtime but when I save the file

### ISSUE THREAD

**Issue by coder3101** [neovim/nvim-lspconfig#1272]
### Description

The diagnostics update is not happening as the code in buffer changes but when I save the changes.

### Neovim version

NVIM v0.6.0-dev+331-gaba397991
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions     -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security     -fstack-clash-protection -fcf-protection -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/ashar/.cache/yay/neovim-git/src/build/config -I/home/ashar/.cache/yay/neovim-git/src/neovim-git/src -I/usr/include -I/home/ashar/.cache/yay/neovim-git/src/build/src/nvim/auto -I/home/ashar/.cache/yay/neovim-git/src/build/include
Compiled by ashar@arch

### Nvim-lspconfig version

Latest as of now

### Operating system and version

Arch Linux Updated

### Affected language servers

metals

### Steps to reproduce

Install as mentioned in `config.md` of `nvim_lspconfig`

### Actual behavior

When I hit `:w` then diagnostics are updated.

### Expected behavior

Diagonistics should update as soon as leave insert mode. I have set it up like that and it works for other languages.

### Minimal config

```Lua
Customized publish diagnostics 

vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    virtual_text = true,
    signs = true,
    update_in_insert = false,
  }
)

```

Setup for all LSPs
```
local servers = { "clangd", "cmake", "metals", "pylsp", "rls", "tsserver", "yamlls", "jsonls"}
for _, lsp in ipairs(servers) do
  nvim_lsp[lsp].setup { on_attach = on_attach }
end
```

`on_attach`
```
function(client, bufnr)
  local function buf_set_keymap(...) vim.api.nvim_buf_set_keymap(bufnr, ...) end
  require('completion').on_attach(client)
  print('Language Server attached')

  -- Mappings.
  local opts = { noremap=true, silent=true }
  buf_set_keymap('n', '<leader>gD', '<cmd>lua vim.lsp.buf.declaration()<CR>', opts)
  buf_set_keymap('n', '<leader>gd', '<cmd>lua vim.lsp.buf.definition()<CR>', opts)
  buf_set_keymap('n', '<leader>gi', '<cmd>lua vim.lsp.buf.implementation()<CR>', opts)
  buf_set_keymap('n', '<leader>gS', '<cmd>lua vim.lsp.buf.signature_help()<CR>', opts)
  buf_set_keymap('n', '<leader>gt', '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)
  buf_set_keymap('n', '<leader>gr', '<cmd>lua vim.lsp.buf.rename()<CR>', opts)
  buf_set_keymap('n', '<leader>gR', '<cmd>lua vim.lsp.buf.references()<CR>', opts)
  buf_set_keymap('n', '<leader>n', '<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>', opts)
  buf_set_keymap('n', '<leader>N', '<cmd>lua vim.lsp.diagnostic.goto_next()<CR>', opts)
  buf_set_keymap('n', '<leader>af', '<cmd>lua vim.lsp.buf.code_action()<CR>', opts)
  buf_set_keymap('n', 'K', '<cmd>lua vim.lsp.buf.hover()<CR>', opts)

  -- Set some keybinds conditional on server capabilities
  if client.resolved_capabilities.document_formatting then
    buf_set_keymap("n", "<leader>=", "<cmd>lua vim.lsp.buf.formatting()<CR>", opts)
  elseif client.resolved_capabilities.document_range_formatting then
    buf_set_keymap("v", "<leader>=", "<cmd>lua vim.lsp.buf.range_formatting()<CR>", opts)
  end
end
```
```


### Health check


health#lspconfig#check
========================================================================
## Checking language server protocol configuration
  - INFO: cmake: configuration checked.
  - ERROR: pylsp: The given command "pylsp" is not executable.
  - ERROR: jsonls: The given command "vscode-json-language-server" is not executable.
  - INFO: clangd: configuration checked.
  - INFO: yamlls: configuration checked.
  - INFO: rls: configuration checked.
  - INFO: sumneko_lua: configuration checked.
  - INFO: tsserver: configuration checked.
  - INFO: metals: configuration checked.



### LSP log

https://gist.github.com/coder3101/2964f7145f0e04f98df56d0d63238949

**Comment by mjlbach** [comment]
@ckipp01 Do you have any idea what could be going on here?

**Comment by coder3101** [comment]
It does not works on macOS Big Sur with latest stable neovim nightly as well.

**Comment by ckipp01** [comment]
> The diagnostics update is not happening as the code in buffer changes but when I save the changes.

This is the expected behavior in Metals. There are multiple types of diagnostics that are coming from Metals. The first being more syntactic errors that we can tell really quickly and can give you as you type. For example if you just start typing nonsene you'll see this since we can already tell that syntactically this isn't correct Scala code. These actually don't require a save:

<img width="418" alt="Screenshot 2021-09-23 at 08 40 55" src="https://user-images.githubusercontent.com/13974112/134464436-88f1bfcb-e01d-48b3-abbd-ba80ef9dc551.png">

For all the other diagnostics, and ones that you probably care most about, the save is necessary. The save is what triggers the actual compilation of your code. Behind the scenes Metals is actually using a Build Server that we are communicating with via [BSP](https://build-server-protocol.github.io/). That's also what is forwarding compilation diagnostics to us, and then from us to Neovim.

In short, the save to compile your code is expected.

**Comment by mjlbach** [comment]
Ah, great. Sorry for the ping but glad to here this is expected. We already redirect users to use nvim-metals, but I can add a note about this in our documentation.

**Comment by ckipp01** [comment]
> Ah, great. Sorry for the ping but glad to hear this is expected.

No worries at all. Always feel free to ping me on these.

**Comment by coder3101** [comment]
@ckipp01 is this behaviour specific to neovim or all other clients like vscode, intelliJ have same limitations.

**Comment by ckipp01** [comment]
> @ckipp01 is this behaviour specific to neovim or all other clients like vscode, intelliJ have same limitations.

All LSP clients that Metals works with will respond the same way. So yes, VS Code will also require a save.

IntelliJ isn't an LSP client, but rather a full fledged IDE. So the behavior you'll experience with IntelliJ is quite different. However in IntelliJ a save is not required since it "auto saves" for lack of a better term.
